support letsencrypt via acme-client

bebehei opened the pull request:
Hi all. I'm currently converting my setup from uberspace to a self-hosted server and this is the perfect project. Thank you for this awesome project!

So I thought, there could be a container added to support letsencrypt by default.

It's not fully tested yet, but I'd like to know what you guys think in general about this!

Things to solve:

- [x] regenerate `dhparam.pem` if it's the fakeoil one
- [x] trigger restart of nginx, dovecot and postfix (any service missing?)
- [x] initial account regeneration

Also another question for understanding: What does `ca.pem` do? I don't see, why there is a CA needed and also there is no reference to `ca.pem` in the code.

chriscroome commented:
The `ca.pem` file would be the [Let's Encrypt intermediate cert](https://letsencrypt.org/certificates/#intermediate-certificates).

Personally I have been using [acme.sh](https://github.com/Neilpang/acme.sh) as a Let's Encrypt client, after using the EFF Python client for some time, I haven't tried [dehydrated](https://github.com/lukas2511/dehydrated).

TBK commented on data/Dockerfiles/letsencrypt/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+FROM alpine
```
It is best practice to lock base image version - e.g. alpine:3.6

TBK commented on data/Dockerfiles/letsencrypt/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+    bash \
+    curl \
+    openssl \
+    && rm -rf /var/cache/apk/*
```
Use the --no-cache parameter instead of `&& rm -rf /var/cache/apk/*`

TBK commented on data/Dockerfiles/letsencrypt/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+    openssl \
+    && rm -rf /var/cache/apk/*
+
+RUN curl -L https://github.com/lukas2511/dehydrated/archive/master.tar.gz \
```
Using multiple RUNs adds additional layers, by using && the work is done in one.
`&& curl -L https://github.com/lukas2511/dehydrated/archive/master.tar.gz`

It is also prefered to lock the external depencendy to a specific version:
https://github.com/lukas2511/dehydrated/archive/v0.4.0.zip

The container philosophy is to create an immutable application.

An alternative could also be to use one of the ACME clients that have already been packaged for Alpine:
https://pkgs.alpinelinux.org/package/v3.6/community/x86_64/acme-client (in C)
https://pkgs.alpinelinux.org/package/edge/community/x86_64/lego (in GO)

TBK commented on data/Dockerfiles/letsencrypt/root/etc/dehydrated/config:
```diff
@@ -0,0 +1,7 @@
+CONTACT_EMAIL="${LE_CONTACT_EMAIL:-$CONTACT_EMAIL}"
```
This should just be placed in `data/Dockerfiles/letsencrypt/` - ` data/Dockerfiles/letsencrypt/config`

TBK commented on data/Dockerfiles/letsencrypt/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+RUN curl -L https://github.com/lukas2511/dehydrated/archive/master.tar.gz \
+  | tar -xzf -
+
+ADD root/ /
```
`COPY` is prefered over `ADD`.
https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/#add-or-copy

TBK commented on data/Dockerfiles/letsencrypt/root/srv/docker.sh:
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
```
This should just be placed in `data/Dockerfiles/letsencrypt/`. 
With the `COPY `command you can decide where it is copied to.

TBK commented on data/Dockerfiles/letsencrypt/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+FROM alpine
+
```
Maintainer should be added:
`LABEL maintainer "Andre Peters <andre.peters@servercow.de>"`

bebehei replied:
Question for understanding: Is `LABEL maintainer` equivalent to `MAINTAINER`?

TBK replied:
It is the replacement, MAINTAINER has been deprecated.

bebehei committed:
add acme-client to support certs automatiaclly signed by letsencrypt

bebehei commented on docker-compose.yml:
```diff
@@ -274,6 +274,31 @@ services:
+      build: ./data/Dockerfiles/acme
+      # All domains to be included in the certificate
+      command: ${MAILCOW_HOSTNAME}
+      #autodiscover.${MAILCOW_HOSTNAME} autoconfig.${MAILCOW_HOSTNAME}
```
is https for the autodiscover and autoconfig domain needed?

TBK replied:
According to this reddit thread it has to be TLS encrypted.
https://www.reddit.com/r/exchangeserver/comments/3qm2l8/autodiscoverdomaincom_is_getting_ssl_certificate/

And this MS KB:
https://support.microsoft.com/da-dk/help/2772058/-the-name-on-the-security-certificate-is-invalid-or-does-not-match-the-name-of-the-site-error-in-outlook-in-a-dedicated-or-itar-office-365-environment

bebehei commented on docker-compose.yml:
```diff
@@ -274,6 +274,31 @@ services:
+      command: ${MAILCOW_HOSTNAME}
+      #autodiscover.${MAILCOW_HOSTNAME} autoconfig.${MAILCOW_HOSTNAME}
+      environment:
+        - CONTAINERS_RESTART=mailcowdockerized_postfix-mailcow_1 mailcowdockerized_dovecot-mailcow_1 mailcowdockerized_nginx-mailcow_1
```
Hardcoding `mailcowdockerized` is bad, but neccessary in here. How can we change this?

TBK replied:
Could ${COMPOSE_PROJECT_NAME} from the mailcow.conf file be used?

bebehei replied:
I never heard such a variable `${COMPOSE_PROJECT_NAME}`. Where should this be defined? Nothing in `mailcow.conf`.

TBK replied:
https://github.com/mailcow/mailcow-dockerized/blob/dev/generate_config.sh

bebehei replied:
I see. But I'm merging against `master` 😕

bebehei commented on docker-compose.yml:
```diff
@@ -274,6 +274,31 @@ services:
+      dns_search: mailcow-network
+      # TODO: find some way to use always
+      # do not restart the container after exit. Things get worse when we hit let's encrypt's ratelimit.
+      restart: "no"
```
Is there any better restart-policy than `"no"`? Using `always` will result in hitting the ratelimit. This is much worse than a stopped container. If hit, the ratelimit cannot get reset manually and requires hours of time to wait. A stopped container can get started and the problem is fixed in seconds.

TBK replied:
There is also `on-failure:`

![image](https://user-images.githubusercontent.com/858296/26901043-af434ef6-4bd4-11e7-8f63-f8d11de2381e.png)

https://docs.docker.com/engine/reference/run/#restart-policies-restart

bebehei replied:
Yeah, I know 'bout that. actually `on-failure` is the bad one.

bebehei replied:
Well, reading the docs again:

> If a container is successfully restarted (the container is started and runs for at least 10 seconds), the delay is reset to its default value of 100 ms.

I always had in my mind, this was at 1s instead of 10s. acme-client should fail in less than 10s. So this is safe to use with `on-failure:3` and not hit rate limits.

Thanks.

bebehei commented:
So, I've taken @TBK0000's comments into account and have rewritten it with `acme-client`. It works now out of the box, but some small questions are still open. Could someone of the dev team @mailcow/contributors review it?

TBK commented on data/Dockerfiles/acme/Dockerfile:
```diff
@@ -0,0 +1,12 @@
+
+RUN apk add --update --no-cache \
+      acme-client \
+      docker \
```
I was wondering if wouldn't be possible to achieve the same functionality with API calls (curl) then installing docker wouldn't be necessary - https://docs.docker.com/engine/api/get-started/

bebehei replied:
Yes. docker sucks up 90MB of 100MB in the disk image. I'm aware of that problem. But I just have not known yet another api. Let's have a look. Thanks.

TBK replied:
Here is the API docs - https://docs.docker.com/engine/api/v1.29/#operation/ContainerRestart

bebehei replied:
Well, that's awesome! Thanks for reminding me. curl works like a charm!

bebehei committed:
replace docker and use curl to make API calls

andryyy committed:
Create docker.sh

- Dovecot regenerates the dhparams.
- Remove version from Docker api call
- Set autoconfig und autodiscover, add ADDITIONAL_SAN for additional SAN
- Bash

andryyy committed:
Update docker.sh

Fix MAILCOW_HOSTNAME...

andryyy committed:
Create docker-compose.yml

andryyy committed:
Create Dockerfile

andryyy committed:
Create docker-entrypoint.sh

bebehei commented on data/Dockerfiles/acme/docker-entrypoint.sh:
```diff
@@ -18,18 +18,14 @@ while true; do
 
-			# also generate new dhparams with new certs
-			#(on the first run, the old dhparams may be the fakeoil from git)
-			openssl dhparam -out ${ACME_BASE}/dhparams.pem 2048
```
@andryyy Could you please elaborate this?

In contrast to `key.pem`, `dhparams.pem` can be public. So it's ok to use the ones from the git-repository. But is it desirable to use the ones from the git-repository?

I see this PR actually to enable SSL on a zero effort basis. So I don't expect anyone doing manual work for SSL while using this container.

So what do you think: Is it desirable to have every mailcow user running with the same dhparams?

bebehei replied:
@andryyy I've seen the explanation in the commit message about dhparams. Sorry for the noise.

andryyy replied:
No problem! :)

bebehei replied:
But... why does dovecot regenerate `dhparams.pem` while it has snakeoil certs? So if dhparams get regenerated, but not the certs, ssl will still be broken (dovecot, nginx **and postfix**). What's the point on regenerating dhparams without knowing if SSL termination will be successful?

I can hear mailbox.org loudly crying about the majority of mailservers having only broken SSL.

bebehei commented on docker-compose.yml:
```diff
@@ -280,9 +280,10 @@ services:
       environment:
         - CONTAINERS_RESTART=mailcowdockerized_postfix-mailcow_1 mailcowdockerized_dovecot-mailcow_1 mailcowdockerized_nginx-mailcow_1
+        # Additional subject alternate names
+        - ADDITIONAL_SAN=
```
What about moving these into mailcow.conf?

andryyy replied:
Yes, good idea.

andryyy commented:
It is missing any reference to nginx-mailcow to actually respond  to the challenge. :o 

//Edit: Forget it. Sorry. :)

Wow, starting and stopping the services really makes you hit the rate limit very fast.

bebehei commented:
Testing at first with the `-s` option in acme-client will give you more time. During development of this PR, I've also hit the RL quite often.

Also I have to say, it works like a charm on my machine.

Dragnucs commented on data/Dockerfiles/acme/docker-entrypoint.sh:
```diff
@@ -0,0 +1,43 @@
+	for container in $*; do
+		curl -X POST \
+			--unix-socket /var/run/docker.sock \
+			"http:/containers/${container}/restart"
```
Shouldn't this strart with http://containers?

bebehei replied:
Well, this was my first thought too. But it's consistently written in the [docker documentation](https://docs.docker.com/engine/api/get-started).

bebehei replied:
Well, I'm currently thinking about this. But I'm sure, that the `containers`-bit is part of the filename in the URL and not some server name. And a possible explanation is that the `//example.com` in an URL is an optional hostname. Because we're using `--unix-socket` to connect to docker, there is no hostname needed and therefore we're using the "default" hostname.

If this is true. That 'd be awesome.

RealOrangeOne replied:
I believe you need a double slash in the protocol? `http://`

andryyy replied:
It works like this. :) I thought it was incorrect, too.

lazyfrosch replied:
Wild guess: http URL without a hostname (since its a unix socket)

bebehei commented:
@andryyy What's the status of this PR? Any info when you will merge it?

andryyy commented:
@bebehei Waiting for Rspamds 1.6.0 release (will be released within the next hours). Then we can finally drop Rmilter, merge the UI changes and have ARC support in mailcow! :)
Thanks to @vstakhov for his awesome work!

bebehei commented:
> ARC

What's this?

TBK commented:
Rspamd 1.6.0 has been release.

TBK commented:
@bebehei https://rspamd.com/doc/modules/arc.html

mkuron commented:
Looks like @andryyy already pushed part of this PR to the dev branch.

lazyfrosch commented on data/Dockerfiles/acme/docker-entrypoint.sh:
```diff
@@ -0,0 +1,43 @@
+		-f ${ACME_BASE}/acme/private/account.key \
+		-k ${ACME_BASE}/acme/private/privkey.pem \
+		-c ${ACME_BASE}/acme \
+		${MAILCOW_HOSTNAME} autoconfig.${MAILCOW_HOSTNAME#*.} autoconfig.${MAILCOW_HOSTNAME#*.} ${ADDITIONAL_SAN}
```
It would be nice to be able to configure the full list. So you don't have to use `autoconfig`

Also the list ist wrong, it should be `autodiscover.${MAILCOW_HOSTNAME#*.} autoconfig.${MAILCOW_HOSTNAME#*.}` --> `discover` :wink:

andryyy replied:
It is autoconfig for Thunderbird and autodiscover for Outlook. :)

lazyfrosch replied:
Yes, but you have listed `autoconfig` twice.

Also autodiscover can be done via an SRV record (what I want to test and would prefer - since it would work better for multiple domains)

Have to test this for autoconfig, but I bet there is SRV support as well...

andryyy replied:
Oh. Didn't notice this. The dev differs a bit though: https://github.com/mailcow/mailcow-dockerized/blob/dev/data/Dockerfiles/acme/docker-entrypoint.sh

This code checks for autodiscover/autoconfig.{each_domain} and tries to create a certificate.

Yes, a SRV works fine, too. Should be part of the docs.

andryyy commented:
I added some more code and am testing it. Looks good so far!

mkuron commented:
Can this be closed? The version that's already in master works quite nicely.

skupfer commented:
Before opening an issue for this question I would like to use this PR...
The guide states mailcow must be running on port 80... Therefore one cannot use this running mailcow on 8080/8443/reverse proxy, right? Could this be achieved in the future?

andryyy commented:
It should work when the RP is available on port 80.
